### PR TITLE
ambient: avoid slow lookup for network

### PR DESCRIPTION
**Pull request by howardjohn**
Today, the ambient hot-path is the workload constructors. These are the
highest scale + highest churn generally. Right now, we call
LookupNetworkGateway each time on this path. This call is not something
that was originally meant to be on a hot path, and is instead cached in
PushContext. Its very slow! We then use this and do an O(n) search which
is, again, pretty slow.

In some testing I was doing, while I was on a single cluster I had a
bunch of networks setup from previously, I found 30% of Istiod CPU on
this function. This is very simple to change into a cache O(1) lookup,
which is what I have done in this PR.

### CODE REVIEW

File: pilot/pkg/serviceregistry/kube/controller/ambient/ambientindex.go [hunk 1]
```diff
@@ -110,9 +113,9 @@ type index struct {
 	XDSUpdater      model.XDSUpdater
 	// Network provides a way to lookup which network a given workload is running on
 	Network LookupNetwork
-	// LookupNetworkGateways provides a function to lookup all the known network gateways in the system.
-	LookupNetworkGateways LookupNetworkGateways
-	Flags                 FeatureFlags
+	// LookupNetworkGatewaysRaw provides a function to lookup all the known network gateways in the system.
+	LookupNetworkGatewaysRaw LookupNetworkGateways
```
Comment (keithmattix) [inline pilot/pkg/serviceregistry/kube/controller/ambient/ambientindex.go:117]: `LookupNetworkGatewaysExpensive`?

File: pilot/pkg/serviceregistry/kube/controller/ambient/ambientindex.go [hunk 2]
```diff
@@ -637,9 +641,36 @@ func (a *index) AdditionalPodSubscriptions(
...
+func (a *index) LookupNetworkGateway(id network.ID) []model.NetworkGateway {
+	n := a.networkGateways.Load()
+	if n == nil {
+		return nil
+	}
+	return (*n)[id]
+}
+
+func (a *index) LookupAllNetworkGateway() []model.NetworkGateway {
+	// Since computing the
```
Comment (keithmattix) [inline pilot/pkg/serviceregistry/kube/controller/ambient/ambientindex.go:662]: Broken comment?

Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/ambient/ambientindex.go:662]: an exercise for the reader /s (thanks!)

**Review by hzxuzhonghu — COMMENTED** [review]
Should you add a new method `getNetworkGateways(network.ID)` in pilot/pkg/serviceregistry/kube/controller/network.go

and cache the gateways by networkID to speed up

this can provide a general underlying lib, could benefit future usage

**Comment by howardjohn** [comment]
> Should you add a new method `getNetworkGateways(network.ID)` in pilot/pkg/serviceregistry/kube/controller/network.go
> 
> and cache the gateways by networkID to speed up
> 
> this can provide a general underlying lib, could benefit future usage

currently the only other usage of the network library is push context which also catches it. 

I was thinking about making the network.go just a bit faster, but I thought it would be very high risk and complex - the code there is extremely complicated and stable for many years

**Comment by hzxuzhonghu** [comment]
Yeah, it it not worth refactoring. My advise is to add a method underlying.

**Comment by howardjohn** [comment]
I don't think it can be done without refactoring. The underlying storage is
split in 2 maps and neither stored by network:
https://github.com/istio/istio/blob/d51e747947db40752f70f487c8e8e658d2d214c4/pilot/pkg/serviceregistry/kube/controller/network.go#L61.
So we would need to maintain a new index over those and keep those up to
date, which I doubt I can do without bugs

**Comment by hzxuzhonghu** [comment]
Is that risky introducing a new map? Looks not complex

**Comment by howardjohn** [comment]
a new map that needs to be kept in sync across the 4-5 different places we
update the state.. it's very complex. If we want to change network.go I
would rather https://github.com/istio/istio/pull/54376. Else is complex and
error prone

**Comment by hzxuzhonghu** [comment]
I do think network is mostly about bussiness logic, should be kept above krt.

Will review this implement though, previous comments just my own suggestion

**Review by hzxuzhonghu — APPROVED** [review]

**Pull request merged**